Fix using *.* in wxDir under MSW

vadz opened the pull request:
Using "*.*" as a wildcard is supposed to match everything under MSW, but
ever since the changes of 4daceaacbd (Check that files returned from
wxDir::FindXXX() match the filter., 2013-04-08, see #3432) it only
matched the files with extension because we double-checked the match
returned by the native MSW function (which does handle "*.*" correctly)
using our own wxString::Matches() which doesn't handle it specially.

Fix this by skipping the call to Matches() when "*.*" is used: we know
that this wildcard matches everything, so rechecking the match would be
useless at best, even if it were not actively harmful. And also skip
this call for "*" because calling Matches("*") always succeeds anyhow.

This also fixes the same bug in wxFind{First,Next}File() and any other
code using them such as wx{File,Dir}Ctrl.

Closes #23905.

vadz committed:
Fix using *.* in wxDir under MSW

Using "*.*" as a wildcard is supposed to match everything under MSW, but
ever since the changes of 4daceaacbd (Check that files returned from
wxDir::FindXXX() match the filter., 2013-04-08, see #3432) it only
matched the files with extension because we double-checked the match
returned by the native MSW function (which does handle "*.*" correctly)
using our own wxString::Matches() which doesn't handle it specially.

Fix this by skipping the call to Matches() when "*.*" is used: we know
that this wildcard matches everything, so rechecking the match would be
useless at best, even if it were not actively harmful. And also skip
this call for "*" because calling Matches("*") always succeeds anyhow.

This also fixes the same bug in wxFind{First,Next}File() and any other
code using them such as wx{File,Dir}Ctrl.

Closes #23905.

PBfordev commented on src/msw/dir.cpp:
```diff
@@ -86,7 +86,13 @@ CheckFoundMatch(const FIND_STRUCT* finddata, const wxString& filter)
+    {
+        // And maybe even skip the check entirely if we have a filter that we
+        // know matches everything. This is more than just an optimization, as
+        // "*.*" it wouldn't match the files without extension according to
```
```suggestion
        // "*.*" wouldn't match the files without extension according to
```

vadz committed:
Remove __WINDOWS__ check from a Windows-only file

No real changes, just remove a redundant __WINDOWS__ check.

vadz committed:
Remove ugly macros and useless functions from wxMSW wxDir code

No real changes, just remove trivial functions used to simply return a
member of a struct and also the completely useless PTR_TO_FINDDATA macro
and simply use the variable directly instead.

vadz committed:
Fix matching "x*.*" in wxMSW wxDir too

Recent commit corrected handling of "*.*", which previously didn't match
the files without extensions, but still left handling of patterns such
as "x*.*" broken, because they are also supposed to match all files
starting with "x" under Windows, whether they have an extension or not.

Fix this by using PathMatchSpecEx() shell function which should handle
this correctly and update the unit test to check for this case as well.

vadz committed:
Work around Wine PathMatchSpecEx() bug in the just added test

Skip this test under Wine until the problem is fixed.

See https://bugs.winehq.org/show_bug.cgi?id=55677

vadz committed:
Add an interactive test for PathMatchSpecEx()

This test allows to compare the results of wxString::Matches(),
wxMatchWild() and MSW native function for the same filter.

PBfordev commented:
BTW, what is the reason of using `PatchMatchSpecEx()` instead of `PatchMatchSpec()` here? I can't see any advantage of the former han and the latter can also be used in 3.2 (both are probably broken under Wine).

vadz committed:
Don't use newer PathMatchSpecEx() unnecessarily

Use older and so hopefully compatible with all MinGW versions
PathMatchSpec() instead as we don't use any of the newer function flags
and it seems to behave identically with the default flag.
